scala/lambdas/ingest-validate-generic-ingest-inputs/src/main/scala/uk/gov/nationalarchives/ingestvalidategenericingestinputs/Lambda.scala

```diff
@@ -0,0 +1,261 @@
+          else ""
+
+        for {
+          _ <- if jsonStructureErrorMessage.isEmpty then IO.unit else log(jsonStructureErrorMessage)
+          _ <-
+            if entriesErrorMessage.isEmpty then IO.unit
+            else
+              log(entriesErrorMessage) >> entriesThatFailedValidation.parTraverse { entry =>
+                val entryId = entry("id") match {
+                  case ValidValidationResult(id) => id
+                  case InvalidValidationResult(listOfErrors) =>
+                    val potentialIdValues = listOfErrors.collect { case valueError: ValueError => valueError.valueThatCausedError }
+                    potentialIdValues.headOption.getOrElse("idUnknown")
+                }
+
+                val fieldNamesWithErrors = entry.collect { case (name, value) if value.getClass.getSimpleName == "InvalidValidationResult" => name }.mkString(", ")
+                val mapToSendAsJson = Map("id" -> entryId, "fieldNamesWithErrors" -> fieldNamesWithErrors, "locationOfFile" -> fullLocationOfResultsFile)
+
+                log(mapToSendAsJson.asJson.noSpaces)
+              }
+          allJsonErrors = entriesThatFailedValidation.+:(jsonStructuralErrors)
+          _ <- Stream
+            .eval(IO.pure(allJsonErrors.asJson.noSpaces))
+            .map(s => ByteBuffer.wrap(s.getBytes()))
+            .toPublisherResource
+            .use { pub =>
+              s3Client.upload(bucket, keyOfResultsFile, FlowAdapters.toPublisher(pub))
+            }
+            .map(_ => ())
+
+          _ <- IO.raiseError(JsonValidationException(jsonStructureErrorMessage + entriesErrorMessage))
+        } yield ()
+      }
+
+    } yield StateOutput(input.batchId, input.metadataPackage)
+
+  private def validateAgainstSchema(metadataJson: String) =
+    for {
+      metadataJsonAsUjson <- IO(read(metadataJson).arr.toList)
+      entriesGroupedByType = metadataJsonAsUjson.foldLeft(Map[String, List[Value]]()) { (typesGrouped, entry) =>
+        val entryType =
+          Try(entry(typeField).str) match {
+            case Success(expectedType) if List("ArchiveFolder", "ContentFolder", "Asset", "File").contains(expectedType) => expectedType
+            case _                                                                                                       => "UnknownType"
+          }
+
+        val entriesBelongingToType = typesGrouped.getOrElse(entryType, Nil)
+        typesGrouped + (entryType -> (entry :: entriesBelongingToType))
+      }
+      validateJsonObjects = validateJsonPerType(entriesGroupedByType)
+      fileEntries <- validateJsonObjects(File)
+      assetEntries <- validateJsonObjects(Asset)
+      archivedFolderEntries <- validateJsonObjects(ArchiveFolder)
+      contentFolderEntries <- validateJsonObjects(ContentFolder)
+      unknownTypeEntries <- validateJsonObjects(UnknownType)
+
+      allEntriesValidatedAndNot = Map(
```

Comment:
Ah, yes, good point! I think I named it like this because initially I was not validating the `UnknownType` ones. But yes, `allValidatedEntries ` sounds good.